### DISCUSSION TITLE
x264, openh264, nvenc, avc444, egfx

### DISCUSSION THREAD

**Discussion by nedix** [neutrinolabs/xrdp#3270] [General]
A few weeks ago I started development of a [Docker container for Kodi using xrdp](https://github.com/nedix/kodi-docker). I compared many screen sharing solutions and xrdp is a clear winner! My goal is to be able to run the container on [vast.ai](https://vast.ai), a marketplace for GPU equiped ([Nvidia Container Toolkit](https://github.com/NVIDIA/nvidia-container-toolkit)) servers.

I have been getting fantastic results on my aarch64 machine with xrdp v0.10.1 and xorgxrdp v0.10.2 using the RFX codec but came just a few fps short for a perfect movie experience. My Samsung TV has RDP support but using the RFX codec and at 1080p this is not running well.

Looking into xrdp I discovered a lot of work has already gone into new codecs, egfx and multimon (I won't be able to test multi monitor however). So I set out to execute the following git operations:

1. Rebased development onto v0.10.1
1. Merged branches from @jsorg71 and @Nexarian
1. Fixed the conflicts so that the project would compile

This has resulted in a fantastic mess that only works with the RFX codec but should contain all the bits and pieces (given some refactoring) to add support for x264, openh264, nvenc, avc444, egfx and multimon (which I won't be able to test) onto the develop branch.

I have set myself a time limit of one week to achieve the following goals:
1. Fix x264 (which broke during the merge operation)
    - Clang Static Analyzer
    - Cppcheck
    - Valgrind
1. AVC444
1. (Nvenc)
1. Clean and prepare PR to develop

If I manage to accomplish my goals within the week I will spend the remaining time on openh264, egfx and multimon. If I run out of time I will extract those bits into feature branches to pick up work some other time (or by somebody else).

<hr>

I very likely will have questions along the way (this is my first C project). I am hoping some of you could subscribe to this discussion so that if I get stuck on something I can ask my questions here. Progress can be followed [over here](https://github.com/nedix/xrdp-fork/tree/main) and in the feature branches. The main branch is protected and requires a PR for merging if you would like to contribute something. Expect a great deal of refactoring in the coming days however.

**Comment by matt335672** [comment]
Dag @nedix 

Sorry for not commenting earlier - I was away last week and only logged back in to Github yesterday.

There has been some work on x.264 in particular on the devel branch, rather than v0.10.x. I think this is worth a look, as a few test cases have been added to that. It may be a better place to start, with AVC444 being added on top later. See #3141 for a bit of background.

Hope this helps.

**Reply by metalefty** [reply]
Also, [v0.10-h264 branch](https://github.com/neutrinolabs/xrdp/tree/v0.10-h264) WIP to merge H.264 into v0.10.

**Comment by nedix** [comment]
Thank you for your replies @matt335672 and @metalefty.

I got to spend about two days last week and then the RPM Fusion repo messed something up which made the Kodi container that I am using to test XRDP uninstallable. The RPM Fusion situation is good now but I had CI/CD plans for this week. I will resume the operation shortly.

Thought I'd share that I've been feeding [CodeChecker](https://codechecker.readthedocs.io) reports into an LLM called Cohere Command-R-Plus (using the Continue plugin for my IDE). This is the first time I am adopting AI for coding and think it provides amazing value for learning about the project and new to me programming language, highly recommend.
